content/things/rust-to-webassembly.md

```diff
@@ -0,0 +1,637 @@
+            19 ┊     0.00% ┊ custom section '.debug_pubtypes' headers
+            18 ┊     0.00% ┊ custom section '.debug_ranges' headers
+            17 ┊     0.00% ┊ custom section '.debug_abbrev' headers
+            16 ┊     0.00% ┊ custom section '.debug_info' headers
+            16 ┊     0.00% ┊ custom section '.debug_line' headers
+            15 ┊     0.00% ┊ custom section '.debug_str' headers
+            14 ┊     0.00% ┊ export "__heap_base"
+            13 ┊     0.00% ┊ export "__data_end"
+            12 ┊     0.00% ┊ custom section 'producers' headers
+             9 ┊     0.00% ┊ export "memory"
+             9 ┊     0.00% ┊ add
+...
+```
+
+It’s now clearly visible that all main contributors to the module size are custom sections, which — by definition — are not relevant for the execution of the module. Their names imply that they contain information that is used for debugging, so it is somewhat surprising they are even emitted for a `--release` build. I suspect that release builds make `rustc` turn on code optimization, remove asserts and other checks. It does not affect whether debug symbols are packed into the binary.
+
+To address this we can use `wasm-strip`, provided yet again by wabt. It’s a tool that removes everything that is unnecessary, including custom sections. After stripping, we are left with a module of a whopping 116B. Disassembling it shows that the only function in that module is called `add` and executes `(f64.add (local.get 0) (local.get 1))`, which means the Rust compiler was able to emit optimal code. Of course, staying on top of binary size gets more complicated with a growing code base.
+
+### Sneaky bloat
+
+I have seen a couple of complaints online about how big WebAssembly modules create by Rust are that do a seemingly small job. In my experience, there are three reasons why Rust emits surprisingly big WebAssembly binaries:
+
+* Debug build (i.e. forgetting to pass `--release` to `Cargo`)
+* Debug symbols (i.e. forgetting to run `wasm-strip`)
+* String formatting and panics 
+
+We have looked at the first two. Let’s take a closer look at the last one. This innocuous line of code compiles to 18KB of WebAssembly:
+
+```rust
+static PRIMES: &[i32] = &[2, 3, 5, 7, 11, 13, 17, 19, 23];
+
+#[no_mangle]
+fn nth_prime(n: usize) -> i32 {
+    PRIMES[n]
+}
+```
+
+Okay, maybe not so innocuous after all. You might already know what's going on here. 
+
+### Panicking
+
+A quick look at `twiggy` shows that the main contributors to the Wasm module size are functions related to string formatting and panicking (before LTO, there was also code for memory management in there). And that makes sense! The parameter `n` is unsanitized and used to index an array. Rust has no choice but to inject bounds checks. If a bounds check fails, Rust panics.
+
+One way to handle this is to do the bounds checking ourselves. Rust's compiler is really good at proving whether undefined behavior can happen or not.
```

Comment:
Hmm "undefined behaviour" is the wrong term to use here, Rust doesn't (can't?) check for it. Maybe you meant "proving whether runtime bound checks are necessary" or something along those lines?